In [1]:
import numpy as np
import pandas as pd 
import matplotlib.pyplot as plt 
%matplotlib inline

### 1. One-Sample T test(평균비교-일표본 T검정)
    => Testing a population mean(모집단의 평균을 테스트해본다)

#### 예제1) :
    => a라는 사람이 매일 식빵 구매 
    식빵 무게 500g 맞는지 확인 하려고 매일 무게 기록 
    495, 496, 502, 494, 499, 501, 493 
    평균 497.1g
    표준편차 3.53g
    평균이 497.1이므로 500g보다 작다 
    단순 평균비교로 컴플레인 거는것보다는 
    통계의 가설검정이 필요 

### 통계적 가설검정 진행 4단계 
    1. 귀무가설 대립가설 설정 
    2. 검정 통계량 계산 
    3. 검정 통계량의 p-value 확인 
    4. p-value와 유의 수준 앞라를 비교하여 판정을 내린다 
        ==> p-value > 유의수준 a ==> 귀무가설 수용 
        ==> p-value < 유의수준 a ===> 귀무가설 기각, 대립가설 수용

### 1) 귀무가설, 대립 가설 설정 
    A는 식빵의 무게가 500g보다 작다고 의심 
    
    H1(대립가설) mean <500 
    H0(귀무가설) mean >=500
    

### 2) 검정 통계량 계산 
t값 공식은??\
T = (mean- 500) / (std / np.sqrt(n))

### 3) 검정 통계량의 p-value 확인 
우리 가설 세울때 빵은 500보다 작다는 가설\
==> 이것은 작다 크다의 개념이므로\
==> 그렇기에 단측 검정(one-tailed test)가 된다 
==> 만약에 같다 같지 않다의 개념이면 양측 검정(two-tailed test)가 된다

In [13]:
import numpy as np
import pandas as pd 
import matplotlib.pyplot as plt 


In [17]:
from scipy.stats import ttest_1samp

In [20]:
from scipy import stats

In [10]:
sample = [495, 496, 502, 494, 499, 501, 493]

In [19]:
ttest_1samp(sample,500)

Ttest_1sampResult(statistic=-2.140129125019247, pvalue=0.07613679645642618)

### 위에서 pvalue는 유의수준a에 따른 귀무가설 기각 판정여부 값으로 
==> 유의수준 a=0.05일 경우 pvalue가 0.05보다 작으면 귀무가설 기각 
==> pvalue가 유의수준 0.05보다 크다면 귀무가설 수용

In [11]:
sample = np.array(sample)

In [12]:
sample.mean(), sample.std()

(497.14285714285717, 3.270149469217028)

In [14]:
t = (sample.mean() - 500) / (sample.std()/np.sqrt(sample.size))

In [15]:
t

-2.3116036533933726

In [23]:
#유의 확률 계산 
p = 1- stats.t.cdf(t,3)

### Two sample T-test
==> 두 독립표본 평균 검정

In [25]:
sample_b = np.array([8.8,8.4,7.9,8.7,9.1,9.6])
sample_g = np.array([9.9,9.0,11.1,9.6,8.7,10.4,9.5])

## 독립된 두 표본에서 추출한 그룹의 평균차이를 검정하려면 
=====> 두 그룹의 분산이 동일하다는 가설을 검정한 이후에\
검정 결과에 따라 다르게 독립표본 t검정을 시행한다 


In [26]:
# 분산 동일성 검정 시행
sample_res2 = stats.levene(sample_b, sample_g)

In [27]:
sample_res2

LeveneResult(statistic=0.5208519677996435, pvalue=0.485536773334305)

#### p값이 통상적인 기준인 0.1보다 크므로 
==>두 그룹의 분산이 동일하다

#### 분산이 동일하다는 가정 받아들여졌으므로 

==> 독립표본 t-검정 시행 

In [28]:
# stats의 ttest_ind()이용, 
# 두 그룹의 분산이 동일하므로 equal_var=True
ttest_res = stats.ttest_ind(sample_b,sample_g,equal_var=True)

In [29]:
ttest_res

Ttest_indResult(statistic=-2.476489813980959, pvalue=0.030764898866015765)

### 검정통계량 t-value가 -2.476, p-value는 0.031
==> p값이 유의수준0.05보다 작으므로 
귀무가설 기각 ==> 두 지혈제간의 효과차이가 없다는 귀무가설 기각\
==> 대립가설 선택(두 지혈제 간의 효과에는 유의한 차이가 있다)

### 예제3: 짝지어진 표본 평균 검정(paired T-test)
아래자료는 10명의 고혈압 환자에게 혈압 강하제 투여 전 후의 혈압을 측정. \
혈압 강하제의 효과가 있었는지 검정해보자

In [30]:
sample1 = np.array([142,140,144,144,142,146,149,150,142,148])
sample2 = np.array([138,136,147,139,143,141,143,145,136,146])

In [31]:
# 대응 표본의 검정에는 stats모듈의 ttest_rel()

pttest_res = stats.ttest_rel(sample1, sample2)

In [32]:
pttest_res

Ttest_relResult(statistic=3.413793103448276, pvalue=0.007703223347263766)

### p-value가 0.05보다 작으므로 
==> 대립가설맞다 + ==> 귀무가설 기각 \
즉 대립가설 맞으므로 혈압강하제는 혈압을 낮추는데 효과가 있다고 해석